File: python/leetcode_questions/is_subsequence.py


class Solution:

    # recursive
    def helper(self, s: str, si: int, t: str, ti: int) -> bool:
        if si >= len(s):
            return True
        if ti >= len(t):
            return False

        if s[si] == t[ti]:
            return self.helper(s, si+1, t, ti+1)
        return self.helper(s, si, t, ti + 1)


    def isSubsequence(self, s: str, t: str) -> bool:
        if not s: return True
        if not t: return False

        return self.helper(s,0,t,0)



        

File: C++/leetcode_questions/IsSubSequence.cpp
#include<bits/stdc++.h>


class Solution
{

    /*
    */


public:
    bool isSubsequence(std::string s, std::string t)
    {
        int sl = s.length(), tl = t.length();
        if (sl > tl) return false;

        int cur = 0;

        for (int i=0;i<tl;i++)
        {
            // move cur along s and bound it by the length of s
            if (s[cur] == t[i])
            {
                cur = std::min(sl,cur+1);
            }
        }

        return cur == sl;
    }
};